### PR TITLE
HDDS-5501. Support to upload/read keys from encrypted buckets through S3G

### PR DESCRIPTION
**Pull request by bharatviswa504**
## What changes were proposed in this pull request?

Add support to put/get/mpu from encrypted bucket from S3G.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5501

## How was this patch tested?

```
Manually tested the fix on secure KMS Cluster.

**S3Gatway**
<property><name>ozone.s3g.kerberos.keytab.file</name><value>/tmp/ozone.keytab</value></property>
<property><name>ozone.s3g.kerberos.principal.key</name><value>s3g/_HOST@ROOT.HWX.SITE</value></property>
**Ranger KMS**
<property><name>hadoop.kms.proxyuser.s3g.users</name><value>*</value></property
<property><name>hadoop.kms.proxyuser.s3g.hosts</name><value>*</value></property><property><name>hadoop.kms.proxyuser.s3g.groups</name><value>*</value></property>

Configure  credentials
[root@xx ~]# kinit -kt /var/run/cloudera-scm-agent/process/71-hbase-REGIONSERVER/hbase.keytab hbase/xx@ROOT.HWX.SITE

[root@xx ~]# ozone s3 getsecret --om-service-id=ozone1
awsAccessKey=hbase/xx@ROOT.HWX.SITE
awsSecret=28805795d4118e1aa4acbd70c7f6915384a30af7ac55f942ccc260ba999bcbba

[root@xx ~]# aws  configure
AWS Access Key ID [****************SITE]: hbase/xx@ROOT.HWX.SITE
AWS Secret Access Key [****************cbba]: 28805795d4118e1aa4acbd70c7f6915384a30af7ac55f942ccc260ba999bcbba
Default region name [None]: 
Default output format [None]:
Test
Create encryption key
hadoop key create key1
**Create enc buck**
[root@nightly71x-4 ~]# ozone sh bucket create /s3v/s3encbuck --bucketkey key1
21/07/26 11:23:06 INFO rpc.RpcClient: Creating Bucket: s3v/s3encbuck, with Versioning false and Storage Type set to DISK and Encryption set to true 

**S3 Test**
[root@xx~]# alias s3api='aws s3api --endpoint https://nightly71x-4.nightly71x.root.hwx.site:9879 --no-verify-ssl'

[root@xx ~]# s3api put-object --bucket s3encbuck --key key1 --body /etc/hadoop/conf/ozone-site.xml 
/usr/lib/fence-agents/bundled/botocore/vendored/requests/packages/urllib3/connectionpool.py:768: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
  InsecureRequestWarning)

[root@xx ~]# s3api get-object --bucket s3encbuck --key key1 /tmp/key1
/usr/lib/fence-agents/bundled/botocore/vendored/requests/packages/urllib3/connectionpool.py:768: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
  InsecureRequestWarning)
{
    "AcceptRanges": "bytes", 
    "ContentType": "application/octet-stream", 
    "LastModified": "Mon, 26 Jul 2021 11:24:21 GMT", 
    "ContentLength": 3785, 
    "Expires": "Mon, 26 Jul 2021 11:25:01 GMT", 
    "CacheControl": "no-cache", 
    "Metadata": {}
}

[root@xx ~]# diff /tmp/key1 /etc/hadoop/conf/ozone-site.xml 

[root@xx ~]# s3api list-objects --bucket s3encbuck
/usr/lib/fence-agents/bundled/botocore/vendored/requests/packages/urllib3/connectionpool.py:768: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
  InsecureRequestWarning)
{
    "Contents": [
        {
            "LastModified": "2021-07-26T11:24:21.458Z", 
            "ETag": "2021-07-26T11:24:21.458Z", 
            "StorageClass": "STANDARD", 
            "Key": "key1", 
            "Size": 3785
        }
    ]
}

```
**Shell Test**

```
[root@nightly71x-4 ~]# ozone sh key put /s3v/s3encbuck/key1 /etc/hadoop/conf/ozone-site.xml 
[root@nightly71x-4 ~]# ozone sh key get /s3v/s3encbuck/key1 /tmp/dkey1
[root@nightly71x-4 ~]# diff /tmp/dkey1 /etc/hadoop/conf/ozone-site.xml  
```

Opened a jira https://issues.apache.org/jira/browse/HDDS-5507 to add secure KMS and add tests for put/get/mpu from encrypted bucket using s3g.

### CODE REVIEW

File: hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/Gateway.java [hunk 1]
```diff
@@ -85,4 +91,27 @@ public void stop() throws Exception {
...
+                  + "Principal: {}, keytab: {}",
+              conf.get(OZONE_S3G_KERBEROS_PRINCIPAL_KEY),
+              conf.get(OZONE_S3G_KERBEROS_KEYTAB_FILE_KEY));
+        }
+
+        SecurityUtil.login(conf, OZONE_S3G_KERBEROS_KEYTAB_FILE_KEY,
+            OZONE_S3G_KERBEROS_PRINCIPAL_KEY);
+      } else {
+        throw new AuthenticationException(SecurityUtil.getAuthenticationMethod(
+            conf) + " authentication method not supported. S3 user login "
```
Comment (vivekratnavel) [inline hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/Gateway.java:110]:
```suggestion
            conf) + " authentication method not supported. S3G user login "
```

**Review by vivekratnavel — COMMENTED** [review]
Overall looks good to me. Just a minor comment inline.

Do we plan to add tests after adding s3g keytab to docker?

**Comment by bharatviswa504** [comment]
Thank You @vivekratnavel for the review.

**Review by vivekratnavel — APPROVED** [review]
+1 LGTM

**Pull request merged**